components/builder-depot/src/backend/s3.rs

```diff
@@ -0,0 +1,181 @@
+// Copyright (c) 2016-2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Pkg storage backend variant which uses S3 (or an API compatible clone) for
+//! hart storage.
+//!
+//! Has been tested against AWS S3.
+//!
+//! All packages are stored in a single bucket, using the fully qualified
+//! package ident as the key, prefixed with `pkgs`
+//!
+//! # Configuration
+//!
+//! Currently the S3Handler must be configured with both an access key
+//! ID and a secret access key.
+use std::fs::File;
+use std::io::{BufReader, Read};
+use std::path::{PathBuf, Path};
+use std::str::FromStr;
+
+use hab_core::package::{PackageTarget, Identifiable, PackageIdent};
+
+use rusoto::{Region, credential::StaticProvider, reactor::RequestDispatcher} ;
+use rusoto_s3::{S3, S3Client, GetObjectRequest, PutObjectRequest, ListObjectsRequest};
+use futures::{Future, Stream};
+
+use DepotUtil;
+use config::S3Cfg;
+
+use error::{Result, Error};
+
+pub struct S3Handler {
+    client: S3Client<StaticProvider, RequestDispatcher>,
+    bucket: String,
+}
+
+impl S3Handler {
+    // The S3 Handler struct contains all of the credential
+    // and target information that we should need to perfom
+    // any backend operations
+    pub fn new(config: S3Cfg) -> Self {
+        let region =
+            match config.endpoint {
+                Some(ref value) => {
+                    Region::Custom { name: "minio_s3".to_owned(), endpoint: value.to_string() }
+                },
+                None => Region::from_str(config.aws_region.as_str()).unwrap()
+        };
+        let aws_id = config.aws_key_id.trim().to_owned();
+        let aws_secret = config.aws_secret_key.trim().to_owned();
+        let cred_provider = StaticProvider::new_minimal(aws_id, aws_secret);
+        let client = S3Client::new(RequestDispatcher::default(), cred_provider, region);
+        let bucket = config.s3bucket_name;
+
+        S3Handler {
+            client: client,
+            bucket: bucket.to_string(),
+        }
+    }
+
+    // Helper function for programmatic creation of
+    // the s3 object key
+    pub fn s3_key<T>(&self, ident: &T, target: &PackageTarget) -> PathBuf
+    where
+        T: Identifiable,
+    {
+        Path::new(ident.origin())
+            .join(format!("{}", ident.name()))
```

Comment:
You might want to leverage this here: https://github.com/habitat-sh/core/blob/master/components/core/src/package/ident.rs#L93-L99

Reply:
We should consider key optimization instead of just sticking with the canonical package name - eg, start keys with the reversed release timestamp to avoid AWS partition hotspots. https://minops.com/blog/2015/01/aws-s3-performance-tuning/

Reply:
I'll read this and adjust accordingly.

Reply:
Ok, because of the way we're generating these keys i don't know if we actually need to change this. Right now we generate a pretty long path which effectively serve to group all of our packages into logical blobs of /origin/packagename/version long before we ever get to the hartfile. I think this optimization should be kept in mind in case we start to see any issues, but based on that blog post i think what we're doing is probably totally fine!

Reply:
Apparently it's the first 6 or so characters at the beginning of the key that really matters for the partitioning.. we're probably fine at the current TPS but if we do hit a limit later it's going to be much harder to switch the key scheme.

Reply:
I can switch it if you think its worth it, it'd be better to do this before we migrate, but the blog post seemed to indicate that creating logical groupings with a pathlike prefix to the key would also serve to group objects together on a partition? Unless I'm misunderstanding what they meant by that, that's basically what we're doing now with our origin.

Reply:
I think since most of our operations are going to be GET request we should be fine fronting our bucket with cloudfront https://docs.aws.amazon.com/AmazonS3/latest/dev/request-rate-perf-considerations.html

Reply:
Yeah the section in there on grouping is what i was referring to. I think we're good on this.